Remove index_put from MM embeddings merging

chenxi-yang opened the pull request:
Summary:
Previously, _merge_multimodal_embeddings used `inputs_embeds[is_multimodal]=flattened` to merge the MM embeddings. This index_put operations calls non_zero in pytorch, which forces an additional D2H shape check sync point.
This diff uses masked_scatter_ to bypass the D2H point. The latency changes from 35ms to 0.08ms.

Test Plan:
E2E Results:
Before:
```
QPS:                 0.91
Avg latency:         8.542s
Avg TTFT (client):   4186.02ms
P50 TTFT (client):   4589.01ms
P99 TTFT (client):   6897.92ms
Avg TTIT (client):   21.78ms
P50 TTIT (client):   19.99ms
P99 TTIT (client):   41.25ms
Avg TTFT (server):   5657.19ms
Avg TTIT (server):   130.71ms
Avg prefill len:     22284.20 tokens
P50 prefill len:     22284.00 tokens
P99 prefill len:     22291.00 tokens
Avg decode len:      200.00 tokens
P50 decode len:      200.00 tokens
P99 decode len:      200.00 tokens
```
After:
```
QPS:                 0.94
Avg latency:         8.456s
Avg TTFT (client):   4089.09ms
P50 TTFT (client):   4510.47ms
P99 TTFT (client):   6902.79ms
Avg TTIT (client):   21.83ms
P50 TTIT (client):   19.89ms
P99 TTIT (client):   41.02ms
Avg TTFT (server):   3808.06ms
Avg TTIT (server):   79.76ms
Avg prefill len:     22284.26 tokens
P50 prefill len:     22284.00 tokens
P99 prefill len:     22291.00 tokens
Avg decode len:      200.00 tokens
P50 decode len:      200.00 tokens
P99 decode len:      200.00 tokens
```

Rollback Plan:

Differential Revision: D79405697

ywang96 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1705,7 +1705,12 @@ def execute_model(
         if max_gen_len == 1:
             # No spec decode tokens.
-            valid_sampled_token_ids = sampled_token_ids.tolist()
+            pinned = torch.empty_like(sampled_token_ids, device='cpu', pin_memory=True)
+            transfer_event = torch.cuda.Event()
+            pinned.copy_(sampled_token_ids, non_blocking=True)
+            transfer_event.record()
+            transfer_event.synchronize()
+            valid_sampled_token_ids = pinned.tolist()
```
This change is unrelated and I believe this is accidental - please remove

ywang96 reviewed (commented):
Thanks for your contritbution and I left some comments! Can you do some end-to-end benchmark and show the overall improvement? Thanks!

shinoss reviewed (commented):
could you fix the pre-commit errors? you can run it locally with `pre-commit run --show-diff-on-failure --color=always --all-files --hook-stage manual `

changes look good.

DarkLight1337 commented on vllm/model_executor/models/utils.py:
```diff
@@ -418,17 +418,22 @@ def _merge_multimodal_embeddings(
+
+        if flattened.shape[0] != num_expected_tokens:
+            expr = _embedding_count_expression(multimodal_embeddings)
+            raise ValueError(
```
Please fix pre-commit

chenxi-yang committed:
Remove index_put from MM embeddings merging (#22105)

Summary:
Pull Request resolved: https://github.com/vllm-project/vllm/pull/22105

Previously, _merge_multimodal_embeddings used `inputs_embeds[is_multimodal]=flattened` to merge the MM embeddings. This index_put operations calls non_zero in pytorch, which forces an additional D2H shape check sync point.
This diff uses masked_scatter_ to bypass the D2H point. The latency changes from 35ms to 0.08ms.

Test Plan:
E2E Results:
Before:
QPS:                 0.91
Avg latency:         8.542s
Avg TTFT (client):   4186.02ms
P50 TTFT (client):   4589.01ms
P99 TTFT (client):   6897.92ms
Avg TTIT (client):   21.78ms
P50 TTIT (client):   19.99ms
P99 TTIT (client):   41.25ms
Avg TTFT (server):   5657.19ms
Avg TTIT (server):   130.71ms
Avg prefill len:     22284.20 tokens
P50 prefill len:     22284.00 tokens
P99 prefill len:     22291.00 tokens
Avg decode len:      200.00 tokens
P50 decode len:      200.00 tokens
P99 decode len:      200.00 tokens
After:
QPS:                 0.94
Avg latency:         8.456s
Avg TTFT (client):   4089.09ms
P50 TTFT (client):   4510.47ms
P99 TTFT (client):   6902.79ms
Avg TTIT (client):   21.83ms
P50 TTIT (client):   19.89ms
P99 TTIT (client):   41.02ms
Avg TTFT (server):   3808.06ms
Avg TTIT (server):   79.76ms
Avg prefill len:     22284.26 tokens
P50 prefill len:     22284.00 tokens
P99 prefill len:     22291.00 tokens
Avg decode len:      200.00 tokens
P50 decode len:      200.00 tokens
P99 decode len:      200.00 tokens

Rollback Plan:

Differential Revision: D79405697

lgeiger commented on vllm/model_executor/models/utils.py:
```diff
@@ -418,17 +418,21 @@ def _merge_multimodal_embeddings(
-            f"Attempted to assign {expr} = {flattened.shape[0]} "
-            f"multimodal tokens to {num_expected_tokens} placeholders")
+    try:
+        inputs_embeds.masked_scatter_(is_multimodal.unsqueeze(-1), flattened)
```
@ywang96 With the upgrade to PyTorch 2.9 this should be handled correctly for index put with a boolean mask as long as the mask is on CPU which at least used to be the case here. See https://github.com/pytorch/pytorch/pull/156384 for more info.
